### DISCUSSION TITLE
AI weather report generation using OpenAI (or ollama) API

### DISCUSSION THREAD

**Discussion by guidocioni** [open-meteo/open-meteo#947] [Show and tell]
Hi all,
I just wanted to share what I think is a cool project that I've been working on in the last few days or so. I wanted to find a way to summarize the weather forecast for a specific day into an easy-to-read text using all data provided by openmeteo. In general we can make our dashboards as pretty as we wanted but, as humans, we'll never be able to analyze automatically and at the same time :)

Turns out this is pretty easy to do using OpenAI API. I just wrote a really long system prompt that instructs the GPT model how to process and interpret the data and to format the output response. Then I just provide the JSON extracted from Openmeteo (with some variables I selected, plus some climatological values and the past forecast to give some context) and the model is able to get me quite a good result.

For example 

>  On August 7, 2024, Hamburg is expected to experience a warm day with maximum
>   temperatures reaching about 26°C. This represents a noticeable increase
>   compared to the previous day's maximum of around 27°C, indicating a warm
>   trend. The nighttime temperature is likely to drop to about 19°C, maintaining
>   a mild evening.
> 
>   Relative humidity will remain relatively high, ranging around 70% during the
>   day, which could contribute to a feeling of warmth. There is a significant
>   risk of precipitation later in the day, particularly in the late afternoon and
>   early evening, with probabilities peaking at around 78%. Total precipitation
>   could reach 0.4 mm, indicating light rainfall.
> 
>   The day will start partially cloudy, evolving into overcast conditions by
>   afternoon, which may aid in keeping temperatures from rising too high. Wind
>   speeds will be moderate at approximately 10 km/h, with gusts potentially
>   reaching up to 41 km/h in the afternoon.
> 
>   Overall, today's weather in Hamburg is set to be warmer than average with a
>   likelihood of scattered rainfall in the latter half of the day.

The funny thing is that I provide many weather variables and the model decide itself what are the most important features to be included in the report. 

I already implemented an undocumented API endpoint in my app (https://github.com/open-meteo/open-meteo/discussions/548) to generate the report, which is then parsed into Home Assistant to give me a notification on my phone.

![Screenshot_20240806-163625](https://github.com/user-attachments/assets/4a6eff7e-ccc4-40af-98c7-539c5c9f241a)

I really cannot wait for the future when the model will become even smarter, but the baseline right now is already quite high. 

As a side note I tried to exchange the Openai API with a local `ollama` server I have, but the open models are definitely not there yet. In all my tests `gpt4-mini` was way more consistent, and the pricing for the moment is not that bad.

**Comment by patrick-zippenfenig** [comment]
Thanks for sharing! Reminds me of https://youtu.be/jzcDi9KJVrk?t=2732. I believe ECMWF is also looking for new projects to build chat bots around weather/climate data

**Reply by guidocioni** [reply]
Yes, it seems quite similar although in that case it seems to be more focused on the technical side of things, that is you want to avoid the user having to retrieve the data himself by writing the code.
In my case I'm more focused on a product that is already processed for general users with forecast data so that they can either get a report like this or ask more direct questions like "what is the likelihood of thunderstorms today at 12".
I'm working on a chatbot enhanced by openmeteo data to integrate directly in the app :)

**Reply by guidocioni** [reply]
<3 

![Screenshot 2024-08-08 at 13 27 15](https://github.com/user-attachments/assets/891402c3-d753-47b6-8e16-f0bd536d99e7)

**Reply by patrick-zippenfenig** [reply]
Thats really quite good... 

Any chance to get this running with open-source LLMs? If you are interested we could setup a GPU server and make a public API available ;-)

**Reply by patrick-zippenfenig** [reply]
Does the new OLLAMA work with 20GB VRAM? https://www.hetzner.com/dedicated-rootserver/matrix-gpu/

**Reply by guidocioni** [reply]
I played around with it and it is indeed amazing... The thing is you can start from really simple question like tell me the general weather forecast, to more specific question like "translate it into german", "what is the uncertainty associated to maximum temperatures" or even "is linen a good fabric to wear tomorrow based on the forecast".
I always got a on-point answer to all of these questions!

And I find even more amazing the fact that you can define the functions to grab the data and expose them to the model so that it will decide when to fetch new data based on the user input. Which means I could just expose all the functions my app currently has to answer more detailed questions like "what is the tmeperature evolution for next week" or "was this year warmer or colder than average up to now"...it opens so many possibilities... 

I do have already an ollama server with openwebui running and I did try to use it instead of the openai API but even trying many different models ranging from 8 to 20B parameters I could never get a relevant answer... Until now openai gpt4 was the only one that was truly working. I wonder if it only has to with the model size. I'm happy to share more details if you're interested :)

**Reply by patrick-zippenfenig** [reply]
That sounds really good. Most weather chat implementation I have seen are terrible.

How expensive are queries to run via the OpenAI API? I am intrigued to make a weather AI chat API publicly available, but this would only work if it would be very cost effective. Running on 1-2 Hetzner GPU servers would be ideal :D. Maybe this could be a project that other national weather services are interested in....

**Reply by guidocioni** [reply]
Our machine has a NVIDIA RTX 4060, Ryzen 9 7950X and 64GB of RAM. We have another machine with a Threadripper 3960X and 128GB with a worse GPU card for more computationally intensive tasks. If you want to experiment with inference you're free to use them (i can give you more details in private).

**Reply by guidocioni** [reply]
I think most weather implementation are bad because
(1) they miss high resolution data, as most people who code them don't have a clue what the best model is for a situation
(2) they miss additional information to give more context: for example it may more beneficial to say that tomorrow is going to be warmer/colder than yesterday instead than giving a plain description of the absolute temperature range

Using the data from openmeteo API and some clever preprocessing of the data I think we can bridge this gap.

**Reply by guidocioni** [reply]
I think the cost is reasonable, although I don't know what it would be in a production environment. Consider that they published the "mini" version of `gpt4o` (which is incidentally the same one I'm using) which slashes cost with a minimum reduction in accuracy. The official pricing is $0.150 / 1M input tokens (https://openai.com/api/pricing/). 
I can tell you I played around with it in the last few days and I only got up to 0.44$ with many test requests

![Screenshot 2024-08-08 at 15 18 08](https://github.com/user-attachments/assets/92a7d99e-1d0a-4623-b6f1-5314144b3ca3)

I think these API become expensive only if you use the "large" models and process a lot of data/audio/images. But as long as it's only JSON and you keep the token number limited (and restrict the scope) it should not be too bad.
